pkg/apis/greenhouse/v1alpha1/types.go

```diff
@@ -48,3 +54,59 @@ type UIApplicationReference struct {
 	// Version of the frontend application.
 	Version string `json:"version"`
 }
+
+// ClusterSelector specifies a selector for clusters by name or by label with the option to exclude specific clusters.
+type ClusterSelector struct {
+	// Name of a single Cluster to select.
+	Name string `json:"clusterName,omitempty"`
+	// LabelSelector is a label query over a set of Clusters.
+	LabelSelector metav1.LabelSelector `json:"labelSelector,omitempty"`
+	// ExcludeList is a list of Cluster names to exclude from LabelSelector query.
+	ExcludeList []string `json:"excludeList,omitempty"`
+}
+
+// ListClusters returns the list of Clusters that match the ClusterSelector's Name or LabelSelector with applied ExcludeList.
+// If the Name or LabelSelector does not return any cluster, an empty ClusterList is returned without error.
+func (cs *ClusterSelector) ListClusters(ctx context.Context, c client.Client, namespace string) (*ClusterList, error) {
+	if cs.Name != "" {
+		cluster := new(Cluster)
+		err := c.Get(ctx, types.NamespacedName{Name: cs.Name, Namespace: namespace}, cluster)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return &ClusterList{}, nil
+			}
+			return nil, err
+		}
+		return &ClusterList{Items: []Cluster{*cluster}}, nil
+	}
+
+	labelSelector, err := metav1.LabelSelectorAsSelector(&cs.LabelSelector)
+	if err != nil {
+		return nil, err
+	}
+	var clusters = new(ClusterList)
+	if err := c.List(ctx, clusters, client.InNamespace(namespace), client.MatchingLabelsSelector{Selector: labelSelector}); err != nil {
+		return nil, err
+	}
```

Comment:
The line already wraps in the Code Review preview, let's do the error handling afterwards.
```suggestion
	err := c.List(ctx, clusters, client.InNamespace(namespace), client.MatchingLabelsSelector{Selector: labelSelector})
	if err != nil {
		return nil, err
	}
```